### PR TITLE
Add closure allocation check to Travis

### PR DESCRIPTION
**Pull request by mihails-strasuns-sociomantic**

### CODE REVIEW

**Comment by mihails-strasuns-sociomantic** [comment]
Ready to go

File: ci/closures.sh [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+export F=devel
+# Enables printing of all potential GC allocation sources to stdout
+export DFLAGS=-vgc
+
+# Prepare docker image and sources
+beaver dlang install
+beaver dlang make d2conv
+
+# Ensure that there are no lines about closure allocations in the output
+! 2>&1 beaver dlang make fasttest | grep "closure"
```
Comment (leandro-lucarella-sociomantic) [inline ci/closures.sh]: Just cosmetic, but the `!` gets a bit lost if you do the redirection at the start, why not
```sh
! beaver dlang make fasttest 2>&1 | grep "closure"
```

**Review by leandro-lucarella-sociomantic — COMMENTED** [review]
I think the script to check should go to beaver though, for re usability purposes

**Comment by leandro-lucarella-sociomantic** [comment]
Mist fail?

File: ci/closures.sh (hunk 1 above)
Comment (mihails-strasuns-sociomantic) [reply ci/closures.sh]: Sure, no preferences here.

**Comment by mihails-strasuns-sociomantic** [comment]
Removed commit used to check that check can indeed fail. Should be green now.

**Comment by mihails-strasuns-sociomantic** [comment]
> I think the script to check should go to beaver though, for re usability purposes

It is pretty small though. I am a bit concerned about adding levels of indirection for 4 line scripts.

**Comment by leandro-lucarella-sociomantic** [comment]
> It is pretty small though. I am a bit concerned about adding levels of indirection for 4 line scripts.

Well, on the other hand if this is copy&pasted to every project and then one adjustment needs to be made, we need to spread it to all projects manually.

**Comment by joseph-wakeling-sociomantic** [comment]
> Well, on the other hand if this is copy&pasted to every project and then one adjustment needs to be made, we need to spread it to all projects manually.

Bear in mind that not all projects use beaver yet, whereas just about all D projects use ocean.  So it's likely more discoverable/useful if it's provided by the latter for now.  (We can always reorganize that later as beaver usage becomes more comprehensive.)

**Comment by nemanja-boric-sociomantic** [comment]
Though the script in ocean already depends on the `beaver` anyway, so we may have it there as well.

**Comment by joseph-wakeling-sociomantic** [comment]
Yea, just noticed :-\  Is there a recommendation for how to integrate this closure check into non-beaver-using libraries and/or applications?

**Comment by nemanja-boric-sociomantic** [comment]
The biggest issue is adapting your CI for it. If you're able to execute arbitrary command on your CI (like in travis), then it's nothing else than:

1. Convert to D2 code
2. Run `!(DFLAGS=-vgc make fasttest 2>&1 | grep -q closure)` or similar

**Comment by mihails-strasuns-sociomantic** [comment]
> grep -q closure

I would actually recommend using it without `-q` - this way you don't need to rerun tests locally to see what caused failure ;)

**Comment by leandro-lucarella-sociomantic** [comment]
> Yea, just noticed :-\ Is there a recommendation for how to integrate this closure check into non-beaver-using libraries and/or applications?

This will be integrated in the new pipeline jobs/library. See https://github.com/sociomantic/krill/pull/799

**Pull request merged**